File: data_prep.R
# This script prepares the data files for processing in Python
# while reducing their size.
rm(list = ls())
library(dplyr)
library(stringr)
complaints_df <- read_csv('/Users/flatironschol/FIS-Projects/Module4/data/NYPD_Complaint_Data_Historic.csv')
offense_types_df <- complaints_df %>% select(KY_CD, OFNS_DESC, PD_CD, PD_DESC) %>% distinct()
borough_df <- complaints_df %>% select(BORO_NM, ADDR_PCT_CD) %>% distinct()
complaints_df <- complaints_df %>% select(CMPLNT_NUM, RPT_DT, ADDR_PCT_CD, KY_CD)
complaints_df <- complaints_df %>% mutate(YEAR = as.numeric(str_sub(RPT_DT,-4,-1)))
complaints_df <- complaints_df %>% select(-RPT_DT) %>% group_by(YEAR, ADDR_PCT_CD, KY_CD) %>% summarise(CMPLNTS = n())
write.csv(complaints_df, file='/Users/flatironschol/FIS-Projects/Module4/data/complaints_df.csv', row.names=FALSE)
write.csv(offense_types_df, file='/Users/flatironschol/FIS-Projects/Module4/data/offense_types_df.csv', row.names=FALSE)
write.csv(borough_df, file='/Users/flatironschol/FIS-Projects/Module4/data/borough_df.csv', row.names=FALSE)

rm(list = ls())
library(dplyr)
library(stringr)
arrests_df <- read_csv('/Users/flatironschol/FIS-Projects/Module4/data/NYPD_Arrests_Data__Historic_.csv')
arrests_df <- arrests_df %>% 
              mutate(YEAR = as.numeric(str_sub(ARREST_DATE,-4,-1)), 
                     MONTH = as.numeric(str_sub(ARREST_DATE,-7,-6))) %>% 
              select(ARREST_KEY, YEAR, MONTH, KY_CD, ARREST_PRECINCT, AGE_GROUP:PERP_RACE)
write.csv(arrests_df, file='/Users/flatironschol/FIS-Projects/Module4/data/arrests_df_ind.csv', row.names=FALSE)
arrests_df <- arrests_df %>%
              group_by(YEAR, ARREST_PRECINCT) %>%
              summarise(ARRESTS = n())
write.csv(arrests_df, file='/Users/flatironschol/FIS-Projects/Module4/FIS-Mod4-Project/data/arrests_df.csv', row.names=FALSE)

rm(list = ls())
library(dplyr)
library(readxl)
sqf_df = data.frame()
for (year in 2016:2018) {
  if (year <= 2014) {
    filename <- paste0('/Users/flatironschol/FIS-Projects/Module4/data/',year,'.csv',collapse='')
    df <- read_csv(filename)
    df <- df %>% mutate_if(is.numeric,as.character) %>% mutate_if(is.logical,as.character) %>% mutate(datestop = as.character(datestop), dob = as.character(dob))
  } else if (year <= 2016) {
    filename <- paste0('/Users/flatironschol/FIS-Projects/Module4/data/sqf-',year,'.csv',collapse='')
    df <- read_csv(filename)
    df <- df %>% mutate_if(is.numeric,as.character) %>% mutate_if(is.logical,as.character)
  } else {
    filename <- paste0('/Users/flatironschol/FIS-Projects/Module4/data/sqf-',year,'.xlsx',collapse='')
    df <- read_excel(filename)
    df <- df %>% mutate_if(is.numeric,as.character) %>% mutate_if(is.logical,as.character)
  }
  sqf_df <- bind_rows(sqf_df, df)
}
sqf_df <- sqf_df %>% 
          mutate(STOP_FRISK_ID = ifelse(is.na(STOP_FRISK_ID), as.numeric(ser_num), as.numeric(STOP_FRISK_ID)), 
                 YEAR2 = ifelse(is.na(YEAR2), as.numeric(year), as.numeric(YEAR2)), 
                 SUSPECT_ARRESTED_FLAG = ifelse(is.na(SUSPECT_ARRESTED_FLAG), arstmade, SUSPECT_ARRESTED_FLAG), 
                 STOP_LOCATION_PRECINCT = ifelse(is.na(STOP_LOCATION_PRECINCT), as.numeric(pct), as.numeric(STOP_LOCATION_PRECINCT))) %>% 
          select(STOP_FRISK_ID, YEAR2, STOP_LOCATION_PRECINCT, SUSPECT_ARRESTED_FLAG)
sqf_df <- sqf_df %>% 
          mutate(SUSPECT_ARRESTED_FLAG = recode(SUSPECT_ARRESTED_FLAG,
                                                "N" = 0,
                                                "Y" = 1)) %>%
          group_by(YEAR2, STOP_LOCATION_PRECINCT) %>%
          summarise(STOPS = n(), STOP_ARRESTS = sum(SUSPECT_ARRESTED_FLAG))
write.csv(sqf_df, file='/Users/flatironschol/FIS-Projects/Module4/FIS-Mod4-Project/data/sqf_df_1.csv', row.names=FALSE)

rm(list = ls())
library(dplyr)
population_df <- read_csv('/Users/flatironschol/FIS-Projects/Module4/data/NYC_Blocks_2010CensusData_Plus_Precincts.csv')
population_df <- population_df %>%
                 select(precinct, P0010001) %>%
                 group_by(precinct) %>%
                 summarise(POPULATION = sum(P0010001))
write.csv(population_df, file='/Users/flatironschol/FIS-Projects/Module4/FIS-Mod4-Project/data/population_df.csv', row.names=FALSE)

File: data_prep.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Dec  2 09:22:12 2019

@author: flatironschol
"""
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
population_list = [8.0233, 7.9844, 7.9396, 7.9043, 7.9087, 7.9458, 7.9911, 8.1937, 8.2927, 8.3835, 8.4586, 8.5211, 8.5825, 8.6154, 8.6227, 8.3987]
basics_df = pd.DataFrame({'year': range(2003,2019), 'population': population_list})
offenses = ['violation-offenses', 'misdemeanor-offenses', \
            'non-seven-major-felony-offenses', 'seven-major-felony-offenses']
for offense in offenses:
    file = f'/Users/flatironschol/FIS-Projects/Module4/FIS-Mod4-Project/data/{offense}-2000-2018.xls'
    df = pd.read_excel(file).T.reset_index()
    if offense == 'violation-offenses':
        df = df[[2,5]]
        df = df.rename(columns = {2: 'year', 5: offense})
    elif offense == 'misdemeanor-offenses':
        df = df[[2,20]]
        df = df.rename(columns = {2: 'year', 20: offense})
    elif offense == 'non-seven-major-felony-offenses':
        df = df[[3,12]]
        df = df.rename(columns = {3: 'year', 12: offense})
    elif offense == 'seven-major-felony-offenses':
        df = df[[3,11]]
        df = df.rename(columns = {3: 'year', 11: offense})
    df = df.drop(index = 0)
    df = df.astype('float64')
    basics_df = basics_df.merge(df, on = 'year', how = 'inner')
for offense in offenses:
    basics_df[offense] = basics_df[offense] / (10 * basics_df.population)
fig, ax1 = plt.subplots(1, 1, figsize = (6, 4))
for offense in offenses:
    sns.lineplot(x = 'year', y = offense, data = basics_df, ax = ax1, legend = 'full')
ax1.set_xticks(range(2003, 2019, 1))
ax1.set_xticklabels(range(2003, 2019, 1), rotation=90)
ax1.set_xlabel('')
ax1.set_ylim(0, 9000)
ax1.set_ylabel('Reported offenses per 100,000 population')
ax1.set_title('Declining Crime Rates in NYC')
plt.legend(offenses)
plt.show()

stops_list = [160851, 313523, 398191, 506491, 472096, 540302, 581168, 601285, 685724, 532911, 191851, 45787, 22565, 12404, 11629, 11008]
innocent_stops_list = [140442, 278933, 352348, 457163, 410936, 474387, 510742, 518849, 605328, 473644, 169252, 37744, 18353, 9394, 7833, 7645]
basics_df = basics_df.merge(pd.DataFrame({'year': range(2003, 2019), 'stops': stops_list}))
basics_df = basics_df.merge(pd.DataFrame({'year': range(2003, 2019), 'innocent_stops': innocent_stops_list}))
basics_df['stops'] = basics_df.stops / (10*basics_df.population)
basics_df['stop_arrests'] = (basics_df.stops) - (basics_df.innocent_stops / (10*basics_df.population))
fig, ax2 = plt.subplots(1, 1, figsize = (6, 4))
sns.lineplot(x = 'year', y = 'stops', data = basics_df, ax = ax2)
sns.lineplot(x = 'year', y = 'stop_arrests', data = basics_df, ax = ax2)
ax2.set_xticks(range(2003, 2019, 1))
ax2.set_xticklabels(range(2003, 2019, 1), rotation=90)
ax2.set_xlabel('')
ax2.set_ylim(0, 9000)
ax2.set_ylabel('Stops per 100,000 population')
ax2.set_title('In 2013, Stop-Question-Frisk Downsized in NYC')
plt.legend(['Stops', 'Stop arrests'])
plt.show()

population_df = pd.read_csv('/Users/flatironschol/FIS-Projects/Module4/FIS-Mod4-Project/data/population_df.csv')
complaints_df = pd.read_csv('/Users/flatironschol/FIS-Projects/Module4/FIS-Mod4-Project/data/complaints_df.csv')
arrests_df = pd.read_csv('/Users/flatironschol/FIS-Projects/Module4/FIS-Mod4-Project/data/arrests_df.csv')
sqf_df = pd.read_csv('/Users/flatironschol/FIS-Projects/Module4/FIS-Mod4-Project/data/sqf_df.csv')
complaints_df = complaints_df.groupby(['YEAR', 'ADDR_PCT_CD']).CMPLNTS.sum()
complaints_df = complaints_df.reset_index()
# merge complaints and population data by precinct (assume same population every year)
df = complaints_df.merge(population_df, how = 'outer',
                         left_on = 'ADDR_PCT_CD',
                         right_on = 'precinct',
                         indicator = True,
                         validate = 'many_to_one')
df = df.drop(df.loc[df.POPULATION == 0].index)
df = df.drop(['precinct', '_merge'], axis = 1)
# merge with arrests data by precinct and year
df = df.merge(arrests_df, how = 'outer',
              left_on = ['ADDR_PCT_CD', 'YEAR'],
              right_on = ['ARREST_PRECINCT', 'YEAR'],
              indicator = True,
              validate = 'one_to_one')
df = df.drop(df.loc[df.CMPLNTS.isna()].index)
df = df.drop(['ARREST_PRECINCT', '_merge'], axis = 1)
# merge with stop-and-frisk data by precinct and year
df = df.merge(sqf_df, how = 'outer',
              left_on = ['ADDR_PCT_CD', 'YEAR'],
              right_on = ['STOP_LOCATION_PRECINCT', 'YEAR2'],
              indicator = True,
              validate = 'one_to_one')
df = df.drop(df.loc[df.YEAR.isna()].index)
df = df.drop(['STOP_LOCATION_PRECINCT', 'YEAR2', '_merge'], axis = 1)

df.to_csv('/Users/flatironschol/FIS-Projects/Module4/FIS-Mod4-Project/data/df.csv')
=